### DISCUSSION TITLE
RFC: Declarative animations

### DISCUSSION THREAD

**Discussion by mattgperry** [motiondivision/motionone#33] [Ideas]
Hi all

Over the past couple weeks I've been working on a declarative animation API for WAAPI and vanilla JS. It might look familiar to you:

```javascript
import { pose } from "motion"
```

So far, it works for single elements but I'm working on inheritance next. I thought that now was a pretty good time to put it to you all and see what you think.

## Basic usage

`pose` can set up declarative animations on one or more elements:

```javascript
pose("button", {
  style: { backgroundColor: "red" }
})
```

Whenever `pose` is run with a new `style`, the elements will automatically animate to their new values:

```javascript
pose("button", {
  style: { backgroundColor: "blue" }
})
```

## Gesture animations

The `pose` function currently supports three gesture animations: `hover`, `press` and `inView`.

```javascript
pose("button", {
  style: { backgroundColor: "blue" },
  inView: { backgroundColor: "red" },
  hover: { scale: 1.2 },
  press: { scale: 0.9 }
})
```

Animation options can be set as defaults via the third argument:

```javascript
pose("button", {
  style: { backgroundColor: "blue" },
  inView: { backgroundColor: "red" },
  hover: { scale: 1.2 },
  press: { scale: 0.9 }
}, {
  duration: 2
})
```

Or they can be passed to a specific pose.

```
style: { backgroundColor: "blue", options: { duration: 2 } },
```

## Events

Posed elements receive new events which can be listened to as normal.

```
element.addEventListener("posecomplete", () => {})
```

I decided on this vs callbacks like `onPoseComplete` passed to `pose` because `pose` can animate multiple elements whereas you'd imagine the event handlers to handle just one.

- `posestart`/`posecomplete` (perhaps this should be `poseend` for consistency)
- `hoverstart`/`hoverend`
- `pressstart`/`pressend`
- `viewenter`/`viewleave`

Hover and press differ from what you might imagine from their underlying pointer events. Hover automatically filters out non-mouse pointers. While `pressend` is triggered from a `pointerup` anywhere in the `window` rather than the element itself (and in future may detect single-touches only)

## Inheritance

This is yet to be implemented but I'm currently working on pose inheritance. This is of course where the orchestration options will be introduced. 

Any of `style`, `hover` etc can be set to the names of arbitrary poses:

```javascript
pose("button", {
  style: "hidden",
  hidden: { opacity: 0 }
}) 
```

When these names change, pose changes will propagate through the DOM.

```javascript
pose("button label", {
  hidden: { scale: 0.5 }
})
```

Like Pose and Framer Motion before it, if children have explicitly set one of these, inheritance will be ignored.

```javascript
pose("button label", {
  style: { backgroundColor: "red" },
  hidden: { scale: 0.5 }
})
```

Options like `staggerChildren` will be able to orchestrate animations amongst nodes.

```javascript
pose("button", {
  style: "hidden",
  hidden: { opacity: 0, options: { staggerChildren: 0.1 } }
}) 
```

## Your thoughts

I'd like your thoughts on this API! I like many aspects of it, for instance it is more declarative than the old vanilla Pose API and relationships between nodes don't need to be explicitly defined (if I can get that working!)

However because `pose` can define animations on multiple elements there's currently no API for setting *just* the style on a specific node. For instance in the button example above you'd have to do something like

```javascript
const button = document.getElementById("#mySpecificButton")
pose(button, {
  style: "visible",
  ...allOtherButtonConfig
})
```

Whereas I wonder if something like

```javascript
button.dataset.pose = "visible"

// Or
import { style } from "motion"

style.set(button, "visible")
```

Might be closer to a better solution.

Your thoughts on this and anything else are apprciated!

**Comment by sbesh91** [comment]
I can see this API being particularly useful for animating a webcomponent.  I think, without shadow dom, your thoughts at the end of this document would become more and more obvious for applications living only in the light dom.

**Comment by okikio** [comment]
@mattgperry I'm not sure if this helps but I found this article recently, and it kinda looks awesome, it may help in making decision on the format to go for, [https://williamjuan027.github.io/angular-animations-explorer/post/advanced/route](https://williamjuan027.github.io/angular-animations-explorer/post/advanced/route#:~:text=style(%7B%20position%3A%20%27relative%27%20%7D)%2C)
